tools/gomod-required-updater/internal/updater/updater.go

```diff
@@ -0,0 +1,225 @@
+package updater
+
+import (
+	"fmt"
+	"log"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
+)
+
+func New(git GitOperator, system SystemOperator, config *Config) *Updater {
+	return &Updater{
+		git:    git,
+		system: system,
+		config: config,
+	}
+}
+
+func (u *Updater) Run() error {
+	log.Printf("Starting update process with remote '%s' and branch '%s'", u.config.RepoRemote, u.config.BranchTrunk)
+
+	absRoot, err := filepath.Abs(u.config.RootPath)
+	if err != nil {
+		return fmt.Errorf("failed to resolve root path: %w", err)
+	}
+
+	// If org modules update is enabled, collect all modules with local replaces
+	if u.config.UpdateOrgModules {
+		modulesToAdd, err := u.findLocalReplaceModules(absRoot)
+		if err != nil {
+			return fmt.Errorf("failed to find local replace modules: %w", err)
+		}
+		if len(modulesToAdd) > 0 {
+			log.Printf("Found %d modules with local replace directives", len(modulesToAdd))
+			u.config.ModulesToUpdate = append(u.config.ModulesToUpdate, modulesToAdd...)
+		}
+
+		rootMod, err := u.findRootModule()
+		if err != nil {
+			return fmt.Errorf("failed to find root module: %w", err)
+		}
+		u.config.ModulesToUpdate = append(u.config.ModulesToUpdate, rootMod)
+	}
+
+	// Get SHA after collecting modules
+	log.Printf("Fetching latest SHA from %s/%s", u.config.RepoRemote, u.config.BranchTrunk)
+	sha, err := u.git.GetSHA(u.config.RepoRemote, u.config.BranchTrunk)
+	if err != nil {
+		return fmt.Errorf("failed to get SHA: %w", err)
+	}
+	log.Printf("Using SHA: %s", sha)
+
+	return u.system.Walk(absRoot, func(path string, isDir bool) error {
```

Comment:
WDYT about respecting `.gitignore`? Traversing something like `node_modules` will add a lot of extra overhead

Reply:
`gomods` does a `filepath.Walk` with: https://github.com/jmank88/gomods/blob/main/gomods.go#L94-L102
And it is essentially instant :shrug: 
So this might be fast enough, but also we could leverage `gomods` directly if we wanted to avoid rewriting this: `gomods -c <this tool>`